# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:

import warnings
warnings.simplefilter('ignore')
import pandas as pd
import sqlite3


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
Sales_connection= sqlite3.connect("go_sales.sqlite")
Sales_connection

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [3]:


newQuery = "SELECT * FROM product"
product = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_type"
product_type = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_line"
product_line = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_staff"
sales_staff = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_branch"
sales_branch = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM retailer_site"
retailer_site = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM country"
country = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_header"
order_header = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_details"
order_details = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM SALES_TARGETData"
target = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM returned_item"
returned_item = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM return_reason"
return_reason = pd.read_sql(newQuery,Sales_connection)

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [4]:
product['PRODUCTION_COST'] = pd.to_numeric(product['PRODUCTION_COST'], errors='coerce')

selected_products = product[(product['PRODUCTION_COST'] > 50.00) & (product['PRODUCTION_COST'] < 100.00)]

selected_products[['PRODUCT_NAME','PRODUCTION_COST']]



PRODUCT_NAME  PRODUCTION_COST
13                Blue Steel Putter            60.00
14            Blue Steel Max Putter            81.80
17              Course Pro Golf Bag            80.00
24                  Hibernator Lite            60.00
25                       Hibernator            86.00
28  Hibernator Self - Inflating Mat            78.55
31              Hibernator Camp Cot            65.33
32     Canyon Mule Climber Backpack            62.50
56          Granite Climbing Helmet            52.86
58            Husky Harness Extreme            55.00
69                   Granite Hammer            57.13
71        TrailChef Deluxe Cook Set            85.11
81                        Polar Ice            73.33
83                     Polar Sports            80.00
84                       Polar Wave            60.00
85                    Polar Extreme            72.50
88                     Edge Extreme            80.00
91                        Seeker 35            79.19
92                        Seeker 50            92.58
93           TrailChef Double Flame            75.00
94                   Seeker Extreme            94.12
97                   Glacier Deluxe            56.25
98                      Glacier GPS            78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [5]:
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')

selected_products = product[(product['MARGIN'] < .20) | (product['MARGIN'] > .60)]

selected_products[['PRODUCT_NAME','MARGIN']]

PRODUCT_NAME  MARGIN
29                 Hibernator Pad    0.17
32   Canyon Mule Climber Backpack    0.17
100             BugShield Natural    0.67
101               BugShield Spray    0.67
102         BugShield Lotion Lite    0.70
103              BugShield Lotion    0.63
105             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [6]:
selected_country = country[(country['CURRENCY_NAME'] == 'francs') ]
selected_country.groupby('COUNTRY')

selected_country['COUNTRY']

0          France
7         Belgium
19    Switzerland
Name: COUNTRY, dtype: object

In [7]:
high_margin_products = product[product['MARGIN'] > .50]
unique_intro_dates = high_margin_products['INTRODUCTION_DATE'].unique()
unique_intro_dates


array(['10-12-2019', '18-12-2019', '27-12-2019', '13-1-2020',
       '15-12-2019', '26-10-2016', '15-2-2011'], dtype=object)

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [8]:
selected_verkoopAfdeling= retailer_site[(retailer_site['ADDRESS2'].notnull())&(retailer_site['REGION'].notnull())]
selected_verkoopAfdeling[['ADDRESS1', 'CITY']]


ADDRESS1           CITY
2            4830 Princeton    Albuquerque
5        1700 George Street  San Francisco
13        1488 Orleans Road     Charleston
28      2103 Ashbury Street    Los Angeles
29           333 Ustaff Way      San Diego
..                      ...            ...
367      10423 Kenwood Road     Cincinnati
370   332 South Main Street        Hanover
371     795 South Jefferson    Springfield
376      702 North Meridian  Oklahoma City
377  1210 Somerville Avenue    Chattanooga

[78 rows x 2 columns]

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [9]:
selected_Landen= country[(country['CURRENCY_NAME']=='new dollar')|(country['CURRENCY_NAME']=='dollars')]
selected_Landen['COUNTRY']


3            Taiwan
6         Australia
14    United States
15           Canada
Name: COUNTRY, dtype: object

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [10]:
selected_vestigingen = retailer_site[(retailer_site['POSTAL_ZONE'].str.startswith('D')) & (retailer_site['ADDRESS2'].notnull())]
selected_vestigingen[['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1       ADDRESS2       CITY
205  Röntgenstraße 90  3. Tür rechts  Frankfurt
351   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [11]:
total_returned_products = returned_item.shape[0]
total_returned_products

706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [12]:
aantal_regios = sales_branch['REGION'].nunique()
aantal_regios

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [13]:
import numpy as np
laagste_marge = product['MARGIN'].min()

hoogste_marge = product['MARGIN'].max()

gemiddelde_marge = round(product['MARGIN'].mean(), 2)
data = {
    'Laagste marge': [laagste_marge],
    'Hoogste marge': [hoogste_marge],
    'Gemiddelde marge': [gemiddelde_marge]
}

marge_overzicht = pd.DataFrame(data)
marge_overzicht


Laagste marge  Hoogste marge  Gemiddelde marge
0           0.17            0.7               0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [14]:
aantal_ontbrekende_adressen = retailer_site['ADDRESS2'].isna().count()
aantal_ontbrekende_adressen
retailer_site


RETAILER_SITE_CODE RETAILER_CODE                          ADDRESS1  \
0                    1            89                1117 Franklin Blvd   
1                   10            92  Suite 1500, 802 Docklands Street   
2                  100            95                    4830 Princeton   
3                  101            95                  6630 Burnet Road   
4                  102           108              2189 Louisville Road   
..                 ...           ...                               ...   
386                 95           106                  932 Ohmey Avenue   
387                 96           106                   438 Kelson Road   
388                 97           106                3410 Selson Avenue   
389                 98           106                 73 Stanley Avenue   
390                 99           106                 37 Raleigh Avenue   

    ADDRESS2           CITY            REGION POSTAL_ZONE COUNTRY_CODE  \
0       None       Winnipeg          Manitoba     R2C 0M5            4   
1       None      Vancouver  British Columbia     V6E 3Z3            4   
2     Unit 2    Albuquerque        New Mexico       87186            3   
3       None         Austin             Texas       78769            3   
4       None   Independence          Kentucky       41051            3   
..       ...            ...               ...         ...          ...   
386     None    Los Angeles        California       90211            3   
387     None    Los Angeles        California       90203            3   
388     None      San Diego        California       92132            3   
389     None  San Francisco        California       94135            3   
390     None  San Francisco        California       94129            3   

    ACTIVE_INDICATOR TRIAL888  
0                  1        T  
1                  1        T  
2                  1        T  
3                  1        T  
4                  0        T  
..               ...      ...  
386                1        T  
387                1        T  
388                1        T  
389                1        T  
390                1        T  

[391 rows x 10 columns]

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [15]:
verkochte_producten_met_korting = order_details[order_details['UNIT_SALE_PRICE'] < order_details['UNIT_PRICE']]

verkochte_producten_met_korting['UNIT_PRICE'] = pd.to_numeric(verkochte_producten_met_korting['UNIT_PRICE'], errors='coerce')

gemiddelde_kostprijs_met_korting = verkochte_producten_met_korting['UNIT_PRICE'].mean()

gemiddelde_kostprijs_met_korting


192.88697889934417

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [16]:
grouped_by_position = sales_staff.groupby('POSITION_EN').size()
aantal_medewerkers_per_functie = grouped_by_position.reset_index(name='Aantal medewerkers')
aantal_medewerkers_per_functie


POSITION_EN  Aantal medewerkers
0                Branch Manager                  19
1              District Manager                   4
2               General Manager                   1
3  Level 1 Sales Representative                  13
4  Level 2 Sales Representative                  39
5  Level 3 Sales Representative                  24
6              Regional Manager                   2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [17]:
groupedBY_Telefoon= sales_staff.groupby('WORK_PHONE').size()
aantal = groupedBY_Telefoon.reset_index(name='Aantal_available')

meer_dan_vier = aantal[aantal['Aantal_available'] > 4]
meer_dan_vier

WORK_PHONE  Aantal_available
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [18]:
netherlands_customers = country[country['COUNTRY'] == 'Netherlands']

netherlands_locations = retailer_site.merge(netherlands_customers, how='inner', left_on='COUNTRY_CODE', right_on='COUNTRY_CODE')

netherlands_locations[['ADDRESS1', 'CITY']]

ADDRESS1          CITY
0                    Beets 36         Beets
1                Perestraat 1      Deventer
2   Laan van Meerdervoort 966      Den Haag
3              Kalfjeslaan 47    Amstelveen
4                Startbaan 11    Amstelveen
5       Noorderspoorsingel 28     Groningen
6               Hogehilweg 13     Amsterdam
7              Spoorstraat 43    Varsseveld
8              Westhavenweg 8     Amsterdam
9               Westzeedijk 4        Muiden
10            Gooiergracht 91         Laren
11            Hoofdstraat 150   Pieterburen
12                Olde Diek 2         Laren
13     Amsterdamsestraatweg 8      Maarssen
14        Krayenhoffstraat 11     Amsterdam
15         Overschiestraat 19     Amsterdam
16              Havenbuurt 12        Marken
17               Duivendaal 1    Wageningen
18              Voorstraat 37  Nieuweschans
19         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [19]:
data_met_Eyewear = product_type[product_type['PRODUCT_TYPE_EN'] == 'Eyewear']['PRODUCT_TYPE_CODE']
eyewear_product_names = product[product['PRODUCT_TYPE_CODE'].isin(data_met_Eyewear)]['PRODUCT_NAME']
eyewear_product_names


80        Polar Sun
81        Polar Ice
83     Polar Sports
84       Polar Wave
85    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [20]:
branch_managers = sales_staff[sales_staff['POSITION_EN'] == 'Branch Manager']
sales_by_manager = pd.merge(branch_managers, order_header, how='inner', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')
sales_with_location = pd.merge(sales_by_manager, retailer_site, how='inner', left_on='RETAILER_SITE_CODE', right_on='RETAILER_SITE_CODE')
result = sales_with_location[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()
result

ADDRESS1 FIRST_NAME LAST_NAME
0  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [21]:
managers = sales_staff[sales_staff['POSITION_EN'].str.contains('Manager')]
sales_by_manager = pd.merge(managers, order_header, how='inner', left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE')
result = sales_by_manager[['POSITION_EN', 'ORDER_DATE']].drop_duplicates()
result

POSITION_EN  ORDER_DATE
0  Branch Manager  2022-09-09
1  Branch Manager  2022-10-31
2  Branch Manager  2022-12-13
3  Branch Manager  2022-03-13
4  Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [22]:
# Stap 1: Bereken totalen van verkochte eenheden per product
total_sold_per_product = order_details.groupby('PRODUCT_CODE')['QUANTITY_SOLD'].sum()

# Stap 2: Filter producten waarvoor meer dan 750 stuks zijn verkocht
high_demand_products = total_sold_per_product[total_sold_per_product > 750].index

# Stap 3: Selecteer namen van producten en bijbehorende producttypen
selected_products = product[product['PRODUCT_CODE'].isin(high_demand_products)][['PRODUCT_CODE', 'PRODUCT_NAME']]
product_types = product_type[['PRODUCT_TYPE_CODE', 'PRODUCT_TYPE_NAME']]

# Stap 4: Combineer informatie in een overzicht
result = pd.merge(selected_products, product_types, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE')[['PRODUCT_NAME', 'PRODUCT_TYPE_NAME']]
result


KeyError: 'PRODUCT_CODE'

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [ ]:
product_merge_40_korting = ((product_merge['UNIT_PRICE'].astype(float) - product_merge['UNIT_SALE_PRICE'].astype(float))/product_merge['UNIT_PRICE'].astype(float)) > 0.4
product_merge.loc[product_merge_40_korting,['PRODUCT_NAME']].drop_duplicates()

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [ ]:
returned_item_merge = pd.merge(returned_item,order_details,left_on='ORDER_DETAIL_CODE',how='left',right_on='ORDER_DETAIL_CODE')
returned_item_90 = (returned_item_merge['RETURN_QUANTITY'].astype(float)/returned_item_merge['QUANTITY'].astype(float)) < 0.9
returned_item_merge.loc[returned_item_90,['RETURN_QUANTITY','QUANTITY','RETURN_REASON_CODE']]

NameError: name 'pd' is not defined

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [ ]:
product.groupby('PRODUCT_TYPE_CODE',as_index=False)['PRODUCT_NUMBER'].count()

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [ ]:
retailer_site.groupby('COUNTRY_CODE',as_index=False)['RETAILER_SITE_CODE'].count()

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [ ]:
amsterdam_sales_staff = sales_staff[sales_staff['SALES_BRANCH_CODE'] == 'Amsterdam']
targets_2006 = target[(target['SALES_YEAR'] == 2006) & (target['SALES_STAFF_CODE'].isin(amsterdam_sales_staff['SALES_STAFF_CODE']))]

Empty DataFrame
Columns: [Id, SALES_STAFF_CODE, SALES_YEAR, SALES_PERIOD, RETAILER_NAME, PRODUCT_NUMBER, SALES_TARGET, RETAILER_CODE, TRIAL888]
Index: []

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [ ]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [ ]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 